### ISSUE TITLE
Message delays when system under load [7920]

### ISSUE THREAD

**Issue by jacobperron** [eProsima/Fast-DDS#1062]
If we send reasonably large data (~1MB) once every few seconds, there are sometimes delays in reading the messages locally when the machine is under load. I've noticed that these delays come in bursts.

## Expected Behavior

Messages should arrive consistently.

## Current Behavior

Messages sometimes take 30-40 seconds to arrive, and typically these delays affect consecutive messages in bursts.

## Steps to Reproduce

1. Source a ROS 2 installation
1. Compile the modified ROS 2 talker/listener demos [here](https://github.com/ros2/demos/tree/jacob/repro_comms_delay).

        mkdir -p /tmp/repro_ws/src
        cd /tmp/repro_ws/src
        git clone https://github.com/ros2/demos.git -b jacob/repro_comms_delay
        cd ..
        colcon build --packages-up-to demo_nodes_cpp

2. Run the talker

        source /tmp/repro_ws/install/setup.bash
        ros2 run demo_nodes_cpp talker

3. Run the listener

        source /tmp/repro_ws/install/setup.bash
        ros2 run demo_nodes_cpp listener

4. Induce load on the machine

        stress --cpu 32

The ROS publisher will send an image message every 3 seconds, and the listener will output the time difference between the time stamp in the message and the time it received the message.

After some time (about a minute), we start see large delays in the arrival of messages to the listener.

## System information

- Fast-RTPS version: 1.9.x
- OS: Ubuntu 18.04
- Network interfaces: lo (127.0.0.1), there's also an ethernet  interface (enp0s25), but I presume it shouldn't be involved since communication is limited to one host.
- ROS 2: latest (Foxy development)

## Additional context

Sending a message every 1 second makes the problem even worse sometimes messages never arrive (even though we are using reliability set to reliable for the publisher and subscription).

I'm not able to reproduce this delayed behavior with other supported RMW implementations (Connext or CycloneDDS).

**Comment by MiguelCompany** [comment]
@jacobperron We have been able to reproduce this problem. We have also checked that using synchronous publishing the problem does not happen. You can check how to enable synchronous publishing [here](https://github.com/ros2/rmw_fastrtps#advance-usage)

Regarding other RMW implementations, we know that CycloneDDS uses synchronous publishing by default, see ros2/rmw_cyclonedds#89. And, IIRC, Connext detects if both participants are in the same host, and will use shared memory in that case.

**Comment by dirk-thomas** [comment]
@MiguelCompany but you still investigating the bug with async and plan to fix it asap, right?

**Comment by MiguelCompany** [comment]
> @MiguelCompany but you still investigating the bug with async and plan to fix it asap, right?

Yes, of course.

**Comment by MiguelCompany** [comment]
@jacobperron @dirk-thomas I tested locally with #1063, and it seems it fixed this. Could you check on your side?

**Comment by jacobperron** [comment]
@MiguelCompany I still the same issue when I try with #1063 :thinking:

**Comment by jacobperron** [comment]
Using the `master` branch, I notice improvements related to delays and dropped messages in another scenario I have (CPU load is not as severe). But with the reproduction steps outlined in this ticket, I don't notice an improvement (qualitatively).

**Comment by MiguelCompany** [comment]
With #1207 and #1209, we see important improvements on the scenario posed in this issue.

@jacobperron Could you check on your side?

**Comment by MiguelCompany** [comment]
@jacobperron Friendly ping.

**Comment by jacobperron** [comment]
@MiguelCompany I've retried the steps described in this ticket with https://github.com/eProsima/Fast-DDS/commits/v2.0.0 (which I believe contains the two referenced PRs). I still see messages delayed on the order of 20 seconds. I've also noticed that the delays are periodic, e.g. here's output from the demo listener (where `diff: X` is difference in sent and received time in seconds):

```
[INFO] [1591745591.965143194] [listener]: I heard an image [sent: 1591745591] [received: 1591745591] diff: 0
[INFO] [1591745621.939019063] [listener]: I heard an image [sent: 1591745594] [received: 1591745621] diff: 27
[INFO] [1591745621.939308358] [listener]: I heard an image [sent: 1591745597] [received: 1591745621] diff: 24
[INFO] [1591745621.939521318] [listener]: I heard an image [sent: 1591745600] [received: 1591745621] diff: 21
[INFO] [1591745621.939725334] [listener]: I heard an image [sent: 1591745603] [received: 1591745621] diff: 18
[INFO] [1591745621.939927991] [listener]: I heard an image [sent: 1591745606] [received: 1591745621] diff: 15
[INFO] [1591745621.940132326] [listener]: I heard an image [sent: 1591745609] [received: 1591745621] diff: 12
[INFO] [1591745621.940333101] [listener]: I heard an image [sent: 1591745612] [received: 1591745621] diff: 9
[INFO] [1591745621.940532214] [listener]: I heard an image [sent: 1591745615] [received: 1591745621] diff: 6
[INFO] [1591745621.940730144] [listener]: I heard an image [sent: 1591745618] [received: 1591745621] diff: 3
[INFO] [1591745621.940927475] [listener]: I heard an image [sent: 1591745621] [received: 1591745621] diff: 0
[INFO] [1591745650.033882475] [listener]: I heard an image [sent: 1591745624] [received: 1591745650] diff: 26
[INFO] [1591745650.034151618] [listener]: I heard an image [sent: 1591745627] [received: 1591745650] diff: 23
[INFO] [1591745650.034350508] [listener]: I heard an image [sent: 1591745630] [received: 1591745650] diff: 20
[INFO] [1591745650.034544657] [listener]: I heard an image [sent: 1591745633] [received: 1591745650] diff: 17
[INFO] [1591745650.034738835] [listener]: I heard an image [sent: 1591745636] [received: 1591745650] diff: 14
[INFO] [1591745650.034930447] [listener]: I heard an image [sent: 1591745640] [received: 1591745650] diff: 10
[INFO] [1591745650.035121700] [listener]: I heard an image [sent: 1591745642] [received: 1591745650] diff: 8
[INFO] [1591745650.035311943] [listener]: I heard an image [sent: 1591745646] [received: 1591745650] diff: 4
[INFO] [1591745650.035501294] [listener]: I heard an image [sent: 1591745648] [received: 1591745650] diff: 2
[INFO] [1591745661.011168615] [listener]: I heard an image [sent: 1591745651] [received: 1591745661] diff: 10
[INFO] [1591745661.011539329] [listener]: I heard an image [sent: 1591745654] [received: 1591745661] diff: 7
[INFO] [1591745661.011802057] [listener]: I heard an image [sent: 1591745657] [received: 1591745661] diff: 4
[INFO] [1591745661.012055984] [listener]: I heard an image [sent: 1591745660] [received: 1591745661] diff: 1
[INFO] [1591745687.897935142] [listener]: I heard an image [sent: 1591745663] [received: 1591745687] diff: 24
[INFO] [1591745687.907892371] [listener]: I heard an image [sent: 1591745666] [received: 1591745687] diff: 21
[INFO] [1591745687.908158967] [listener]: I heard an image [sent: 1591745669] [received: 1591745687] diff: 18
[INFO] [1591745687.908366336] [listener]: I heard an image [sent: 1591745672] [received: 1591745687] diff: 15
[INFO] [1591745687.908563832] [listener]: I heard an image [sent: 1591745675] [received: 1591745687] diff: 12
[INFO] [1591745687.908755537] [listener]: I heard an image [sent: 1591745678] [received: 1591745687] diff: 9
[INFO] [1591745687.908943578] [listener]: I heard an image [sent: 1591745681] [received: 1591745687] diff: 6
[INFO] [1591745687.909181881] [listener]: I heard an image [sent: 1591745684] [received: 1591745687] diff: 3
[INFO] [1591745687.909400245] [listener]: I heard an image [sent: 1591745687] [received: 1591745687] diff: 0
```

**Comment by MiguelCompany** [comment]
@jacobperron We think this has been solved by #1597. Could you check on your side?

**Comment by jacobperron** [comment]
@MiguelCompany The demo appears to be running much smoother now, thanks :bow: 
I tested with the latest `master` branch (3a01da4).
